server/routes/api/protein.py

```diff
@@ -13,22 +13,411 @@
+    '''
+    Given interaction DCID and the id of one of the interactors, return the id of the other one.
+    E.g. 'bio/P53_HUMAN_CBP_HUMAN', 'P53_HUMAN' --> 'CBP_HUMAN'
+    '''
+    interaction_id = _id(interaction_dcid)
+    interactor_id1, interactor_id2 = _interactors(interaction_id).values()
+    if interactor_id1 == interactor_id:
+        return interactor_id2
+    elif interactor_id2 == interactor_id:
+        return interactor_id1
+    raise ValueError(f'{interactor_id} not in {interaction_dcid}')
+
+
+def _reverse_interaction_id(interaction_id):
+    '''
+    Given interaction ID of the form '{protein1 id}_{protein2 id}', return '{protein2 id}_{protein1 id}'.
+    E.g. 'P53_HUMAN_CBP_HUMAN' --> 'CBP_HUMAN_P53_HUMAN'
+    '''
+    interactors = _interactors(interaction_id)
+    return f'{interactors["second_interactor"]}_{interactors["first_interactor"]}'
+
+
+def _reverse_interaction_dcid(interaction_dcid):
+    '''
+    Given interaction DCID of the form 'bio/{protein1 id}_{protein2 id}', return 'bio/{protein2 id}_{protein1 id}'.
+    E.g. 'bio/P53_HUMAN_CBP_HUMAN' --> 'bio/CBP_HUMAN_P53_HUMAN'
+    '''
+    return _dcid(_reverse_interaction_id(_id(interaction_dcid)))
+
+
+def _filter_interaction_dcids(interaction_dcids):
+    '''
+    Given list of interaction DCIDs, remove self-interactions and deduplicate (keep one copy of each set of duplicates).
+    Self-interactions are interactions DCIDs of the form 'bio/{protein id}_{protein id}' (e.g. 'bio/P53_HUMAN_P53_HUMAN')
+    Duplicates include both exact duplicates and reverse duplicates.
+    E.g. 'bio/P53_HUMAN_CBP_HUMAN' is duplicate with both itself and 'bio/CBP_HUMAN_P53_HUMAN'.
+    '''
+    dcid_set = set()
+    uniques = []
+    for interaction_dcid in interaction_dcids:
+        try:
+            protein_id1, protein_id2 = _interactors(interaction_dcid).values()
+        # skip malformatted ids
+        except ValueError:
+            continue
+        # filter out self-interactions
+        if protein_id1 == protein_id2:
+            continue
+        reverse_dcid = _reverse_interaction_dcid(interaction_dcid)
+        if interaction_dcid not in dcid_set:
+            uniques.append(interaction_dcid)
+            dcid_set.update([interaction_dcid, reverse_dcid])
+    return uniques
+
+
+def _symmetrized_scores(scores):
+    '''
+    Given a dict mapping interaction IDs to scores, return a new dict D of the same type such that
+    1) For any interaction ID in D, its reversal is also a key of D
+    2) Any interaction ID in D maps to the same value as its reversal
+    3) If an interaction ID is in scores but its reversal is not, D[interaction ID] = scores[interaction ID]
+    4) If both an interaction ID F and its reversal R are in scores, D[F] = D[R] = max(scores[F], scores[R])
+    '''
+    scores_sym = {}
+    for interaction_id, score in scores.items():
+        reverse_id = _reverse_interaction_id(interaction_id)
```

Comment:
should this be in a try? since reverse_interaction_id can throw errors

Reply:
thanks for catching will do!